### PR TITLE
revert + fix($rootScope): fix potential memory leak when removing scope listeners

### PR DESCRIPTION
**Pull request by jbedard**
This reverts the 1.7 change, cherry-picks the 1.6 version that had no breaking changes, and adds some tests to prevent such a breaking change again in the future.

See https://github.com/angular/angular.js/pull/16293#issuecomment-348859074 for the reason to revert the breaking change.

### CODE REVIEW

File: test/ng/rootScopeSpec.js [hunk 1]
```diff
@@ -2803,6 +2803,45 @@ describe('Scope', function() {
...
+        }
+      });
+
+      $rootScope.$emit('evt');
+      $rootScope.$broadcast('evt');
+      expect(callCount).toBe(6);
+    }));
+
+
+    it('should allow recursive $emit/$broadcast between parent/child', inject(function($rootScope) {
```
Comment (gkalpak) [inline test/ng/rootScopeSpec.js]: I don't think this test demonstrates recursive emitting/broadcasting.

**Review by gkalpak — APPROVED** [review]
One test nit. Otherwise LGTM 👍 
(**EDIT**: As soon as Travis is happy :grin:)

File: test/ng/rootScopeSpec.js (hunk 1 above)
Comment (jbedard) [reply test/ng/rootScopeSpec.js]: How so? It does at least one recursive `$broadcast` + one `$emit` within `$broadcast`...

I'm sure we could think of better or more interesting tests though. Let me know if you have any specific ones you think should be added.

Comment (gkalpak) [reply test/ng/rootScopeSpec.js]: Yeah, it does `root.$broadcast` inside `root.$broadcast` (which we don't care about in this test) and `child.$emit` inside `root.$broadcast`, so in a way it is recursive. But it doesn't test for `root.$broadcast` inside `child.$emit`, for example.

In order to test both `root.$broadcast --> child.$emit` and `child.$emit --> root.$broadcast`, I would change it to something like:

```js
var child = $rootScope.$new();
var calls = '';

$rootScope.$on('evt', function($event, arg0) {
  calls += 'r';  // For "root".
  if (arg0 === 'fromChild') {
    $rootScope.$broadcast('evt', 'fromRoot2');
  }
});

child.$on('evt', function($event, arg0) {
  calls += 'c';  // For "child".
  if (arg0 === 'fromRoot1') {
    child.$emit('evt', 'fromChild');
  }
});

$rootScope.$broadcast('evt', 'fromRoot1');
expect(calls).toBe('rccrrc');
```

Comment (jbedard) [reply test/ng/rootScopeSpec.js]: I updated to use your test

**Comment by Narretz** [comment]
Acording to the guidelines, the commit message should start with "revert: ..." but I don't think that's consequential. Imo, we should support the default git Revert, too. Doesn't matter though.

What I find a bit strange is that the Revert commit comes after the commit with the different fix. I think it's easier to understand if their position is reversed.

**Comment by jbedard** [comment]
I think github is showing them based on the commit date, and the cherry-picked commit has a date in august while the revert is more recent? I'm sure I reverted first, otherwise I would have had a lot of conflicts...

**Comment by petebacondarwin** [comment]
Also, the `changez` tool expects reverts to take the Angular format. It uses it to ensure that we don't include reverted commits in the changelog.

**Comment by jbedard** [comment]
What's the angular format for reverts though?

**Comment by gkalpak** [comment]
For `type(scope): subject` it is `revert: type(scope)' subject`.

**Comment by Narretz** [comment]
Yeah, the commits are correctly ordered when viewed locally. Strange.

The commit message header of the revert commit should look like this:

`revert: fix($rootScope): fix potential memory leak when removing scope listeners`

**Comment by Narretz** [comment]
Merged! Thanks @jbedard

**Pull request merged**